File: clients/client-go/retry429_test.go
package queen

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"sync"
	"testing"
	"time"
)

// Client-side 429/403 handling tests (PLAN_QUEEN_PROXY_CLOUD.md §4/§9,
// blocker B4 -- "client 429/backoff work ... mandatory pre-enforcement").
//
// The proxy error contract under test:
//
//	429  Retry-After: <seconds>  {"error": "...", "code": "rate_limited" | "quota_exceeded"}
//	403                          {"error": "...", "code": "cluster_suspended" | "storage_quota_exceeded"
//	                                                      | "feature_gated" | "forbidden"}
//
// These run against a real local httptest server (no broker), mirroring
// ack_test.go's style.

// planResponse is one canned server response for the test double below.
type planResponse struct {
	status     int
	body       interface{} // encoded as JSON; use []interface{} for a raw array body
	retryAfter string      // Retry-After header value, when non-empty
}

type recordedHit struct {
	method string
	path   string
	atMs   int64
}

func rateLimited429(retryAfter string) planResponse {
	return planResponse{
		status:     http.StatusTooManyRequests,
		body:       map[string]interface{}{"error": "slow down", "code": "rate_limited"},
		retryAfter: retryAfter,
	}
}

func repeatResponse(resp planResponse, n int) []planResponse {
	out := make([]planResponse, n)
	for i := range out {
		out[i] = resp
	}
	return out
}

// newPlanServer starts an httptest server that serves `plan` in request
// order, falling back to `defaultResp` once exhausted. Returns the server
// URL and a hit log (method/path/arrival time) for assertions. The server
// and hit log are safe to read after the test body returns (t.Cleanup runs
// Close before the caller inspects final counts in most tests, but the log
// itself is only appended to from the single httptest handler goroutine at
// a time thanks to the mutex).
func newPlanServer(t *testing.T, plan []planResponse, defaultResp planResponse) (string, *[]recordedHit) {
	t.Helper()
	hits := &[]recordedHit{}
	start := time.Now()
	var mu sync.Mutex
	idx := 0

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		mu.Lock()
		resp := defaultResp
		if idx < len(plan) {
			resp = plan[idx]
		}
		idx++
		*hits = append(*hits, recordedHit{method: r.Method, path: r.URL.Path, atMs: time.Since(start).Milliseconds()})
		mu.Unlock()

		if resp.retryAfter != "" {
			w.Header().Set("Retry-After", resp.retryAfter)
		}
		w.Header().Set("Content-Type", "application/json")
		status := resp.status
		if status == 0 {
			status = http.StatusOK
		}
		w.WriteHeader(status)
		body := resp.body
		if body == nil {
			body = map[string]interface{}{"ok": true}
		}
		_ = json.NewEncoder(w).Encode(body)
	}))
	t.Cleanup(server.Close)
	return server.URL, hits
}

func hitCount(hits *[]recordedHit) int {
	return len(*hits)
}

// ---------------------------------------------------------------------------
// HttpClient (direct) -- the centralized retry429 mechanism itself.
// ---------------------------------------------------------------------------

func TestHttpClient429HonorsRetryAfter(t *testing.T) {
	url, hits := newPlanServer(t, []planResponse{rateLimited429("0")}, planResponse{status: 200, body: map[string]interface{}{"ok": true}})
	client, err := NewHttpClient(ClientConfig{URL: url, Retry429: &Retry429Config{BaseMs: 5, CapMs: 50}})
	if err != nil {
		t.Fatalf("NewHttpClient: %v", err)
	}
	defer client.Close()

	result, err := client.Get(context.Background(), "/x", 0, "")
	if err != nil {
		t.Fatalf("Get returned error: %v", err)
	}
	if result["ok"] != true {
		t.Errorf("unexpected result: %+v", result)
	}
	if hitCount(hits) != 2 {
		t.Errorf("hits = %d, want 2 (one 429 then one success)", hitCount(hits))
	}
}

func TestHttpClient429ExponentialBackoffGrows(t *testing.T) {
	url, hits := newPlanServer(t, repeatResponse(rateLimited429(""), 2), planResponse{status: 200, body: map[string]interface{}{"ok": true}})
	client, err := NewHttpClient(ClientConfig{URL: url, Retry429: &Retry429Config{BaseMs: 20, CapMs: 2000}})
	if err != nil {
		t.Fatalf("NewHttpClient: %v", err)
	}
	defer client.Close()

	if _, err := client.Get(context.Background(), "/x", 0, ""); err != nil {
		t.Fatalf("Get returned error: %v", err)
	}
	h := *hits
	if len(h) != 3 {
		t.Fatalf("hits = %d, want 3", len(h))
	}
	gap1 := h[1].atMs - h[0].atMs
	gap2 := h[2].atMs - h[1].atMs
	if float64(gap2) <= float64(gap1)*1.2 {
		t.Errorf("expected gap2 (%dms) to exceed gap1 (%dms) by backoff growth", gap2, gap1)
	}
}

func TestHttpClient429GivesUpAfterMaxAttempts(t *testing.T) {
	url, hits := newPlanServer(t, nil, rateLimited429(""))
	client, err := NewHttpClient(ClientConfig{URL: url, Retry429: &Retry429Config{MaxAttempts: 3, BaseMs: 1, CapMs: 5}})
	if err != nil {
		t.Fatalf("NewHttpClient: %v", err)
	}
	defer client.Close()

	_, err = client.Post(context.Background(), "/api/v1/push", map[string]interface{}{"items": []interface{}{}})
	if err == nil {
		t.Fatal("expected an error after exhausting retry429.MaxAttempts")
	}
	httpErr, ok := err.(*HTTPError)
	if !ok {
		t.Fatalf("expected *HTTPError, got %T: %v", err, err)
	}
	if httpErr.StatusCode != 429 || httpErr.Code != "rate_limited" {
		t.Errorf("HTTPError = %+v, want status 429 code rate_limited", httpErr)
	}
	if hitCount(hits) != 3 {
		t.Errorf("hits = %d, want exactly maxAttempts (3)", hitCount(hits))
	}
}

func TestHttpClient429DefaultsToTenAttempts(t *testing.T) {
	url, hits := newPlanServer(t, nil, rateLimited429(""))
	client, err := NewHttpClient(ClientConfig{URL: url, Retry429: &Retry429Config{BaseMs: 1, CapMs: 2}})
	if err != nil {
		t.Fatalf("NewHttpClient: %v", err)
	}
	defer client.Close()

	_, err = client.Post(context.Background(), "/api/v1/push", map[string]interface{}{"items": []interface{}{}})
	if err == nil {
		t.Fatal("expected an error")
	}
	if hitCount(hits) != 10 {
		t.Errorf("hits = %d, want 10 (default push maxAttempts)", hitCount(hits))
	}
}

func TestHttpClient429PopRetriesPastPushDefault(t *testing.T) {
	url, hits := newPlanServer(t, repeatResponse(rateLimited429("0"), 14), planResponse{
		status: 200,
		body:   map[string]interface{}{"messages": []interface{}{map[string]interface{}{"transactionId": "m1"}}},
	})
	client, err := NewHttpClient(ClientConfig{URL: url, Retry429: &Retry429Config{BaseMs: 1, CapMs: 5}})
	if err != nil {
		t.Fatalf("NewHttpClient: %v", err)
	}
	defer client.Close()

	result, err := client.Get(context.Background(), "/api/v1/pop", 0, "", WithLongPollRetry())
	if err != nil {
		t.Fatalf("Get returned error: %v", err)
	}
	if result == nil || result["messages"] == nil {
		t.Errorf("unexpected result: %+v", result)
	}
	if hitCount(hits) != 15 {
		t.Errorf("hits = %d, want 15 -- pop must not give up at the push default of 10 attempts", hitCount(hits))
	}
}

func TestHttpClient429ExplicitMaxAttemptsAppliesToPopToo(t *testing.T) {
	url, hits := newPlanServer(t, nil, rateLimited429(""))
	client, err := NewHttpClient(ClientConfig{URL: url, Retry429: &Retry429Config{MaxAttempts: 2, BaseMs: 1, CapMs: 5}})
	if err != nil {
		t.Fatalf("NewHttpClient: %v", err)
	}
	defer client.Close()

	_, err = client.Get(context.Background(), "/api/v1/pop", 0, "", WithLongPollRetry())
	if err == nil {
		t.Fatal("expected an error")
	}
	if hitCount(hits) != 2 {
		t.Errorf("hits = %d, want 2 -- explicit MaxAttempts bounds pop as well as push", hitCount(hits))
	}
}

func TestHttpClient403NeverRetriesAndPreservesCode(t *testing.T) {
	url, hits := newPlanServer(t, nil, planResponse{status: 403, body: map[string]interface{}{"error": "cluster suspended", "code": "cluster_suspended"}})
	client, err := NewHttpClient(ClientConfig{URL: url})
	if err != nil {
		t.Fatalf("NewHttpClient: %v", err)
	}
	defer client.Close()

	_, err = client.Post(context.Background(), "/api/v1/push", map[string]interface{}{"items": []interface{}{}})
	if err == nil {
		t.Fatal("expected an error")
	}
	httpErr, ok := err.(*HTTPError)
	if !ok {
		t.Fatalf("expected *HTTPError, got %T: %v", err, err)
	}
	if httpErr.StatusCode != 403 || httpErr.Code != "cluster_suspended" || !httpErr.IsClusterSuspended() {
		t.Errorf("HTTPError = %+v, want 403/cluster_suspended", httpErr)
	}
	if hitCount(hits) != 1 {
		t.Errorf("hits = %d, want 1 -- 403 must not be retried", hitCount(hits))
	}
}

func TestHttpClient400NotRetried(t *testing.T) {
	url, hits := newPlanServer(t, nil, planResponse{status: 400, body: map[string]interface{}{"error": "bad request"}})
	client, err := NewHttpClient(ClientConfig{URL: url})
	if err != nil {
		t.Fatalf("NewHttpClient: %v", err)
	}
	defer client.Close()

	if _, err := client.Get(context.Background(), "/x", 0, ""); err == nil {
		t.Fatal("expected an error")
	}
	if hitCount(hits) != 1 {
		t.Errorf("hits = %d, want 1", hitCount(hits))
	}
}

// ---------------------------------------------------------------------------
// Queen public API wiring: config plumbing (Retry429) + call-site marking
// (push vs. wait=true pop) actually reach HttpClient.
// ---------------------------------------------------------------------------

func TestQueenPushRetries429ThenSucceeds(t *testing.T) {
	url, hits := newPlanServer(t, []planResponse{rateLimited429("0")}, planResponse{
		status: 200,
		body:   []interface{}{map[string]interface{}{"status": "queued", "transactionId": "tx-1"}},
	})
	client, err := New(ClientConfig{URL: url, Retry429: &Retry429Config{BaseMs: 5, CapMs: 50}})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	defer client.Close(context.Background())

	responses, err := client.Queue("q1").Push(map[string]interface{}{"hello": "world"}).Execute(context.Background())
	if err != nil {
		t.Fatalf("Push failed: %v", err)
	}
	if len(responses) != 1 || responses[0].Status != "queued" {
		t.Errorf("responses = %+v, want a single queued response", responses)
	}
	if hitCount(hits) != 2 {
		t.Errorf("hits = %d, want 2", hitCount(hits))
	}
}

func TestQueenPushSurfacesTerminal403(t *testing.T) {
	url, hits := newPlanServer(t, nil, planResponse{status: 403, body: map[string]interface{}{"error": "over quota", "code": "storage_quota_exceeded"}})
	client, err := New(ClientConfig{URL: url})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	defer client.Close(context.Background())

	_, err = client.Queue("q1").Push(map[string]interface{}{"hello": "world"}).Execute(context.Background())
	if err == nil {
		t.Fatal("expected push to fail")
	}
	var httpErr *HTTPError
	if !errors.As(err, &httpErr) {
		t.Fatalf("expected an *HTTPError in the chain, got %T: %v", err, err)
	}
	if httpErr.StatusCode != 403 || httpErr.Code != "storage_quota_exceeded" {
		t.Errorf("HTTPError = %+v, want 403/storage_quota_exceeded", httpErr)
	}
	if hitCount(hits) != 1 {
		t.Errorf("hits = %d, want 1 -- 403 must not be retried", hitCount(hits))
	}
}

func TestQueenPopRidesOutMore429sThanPushDefault(t *testing.T) {
	url, hits := newPlanServer(t, repeatResponse(rateLimited429("0"), 12), planResponse{
		status: 200,
		body: map[string]interface{}{"messages": []interface{}{
			map[string]interface{}{"transactionId": "tx-1", "partitionId": "p-1", "data": map[string]interface{}{"x": 1}},
		}},
	})
	client, err := New(ClientConfig{URL: url, Retry429: &Retry429Config{BaseMs: 1, CapMs: 5}})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	defer client.Close(context.Background())

	messages, err := client.Queue("q1").Wait(true).Pop(context.Background())
	if err != nil {
		t.Fatalf("Pop failed: %v", err)
	}
	if len(messages) != 1 {
		t.Fatalf("messages = %d, want 1", len(messages))
	}
	if hitCount(hits) != 13 {
		t.Errorf("hits = %d, want 13", hitCount(hits))
	}
}

// ---------------------------------------------------------------------------
// ConsumerManager worker loop: the actual hot-loop bug (B4) this task fixes
// -- Consume() must back off through 429s and stop cleanly on a terminal
// 403 instead of spinning with no delay or hanging indefinitely.
// ---------------------------------------------------------------------------

func TestConsumeBacksOffThrough429AndDeliversMessage(t *testing.T) {
	url, hits := newPlanServer(t, repeatResponse(rateLimited429("0"), 3), planResponse{
		status: 200,
		body: map[string]interface{}{"messages": []interface{}{
			map[string]interface{}{"transactionId": "tx-1", "partitionId": "p-1", "data": map[string]interface{}{"x": 1}},
		}},
	})
	client, err := New(ClientConfig{URL: url, Retry429: &Retry429Config{BaseMs: 1, CapMs: 5}})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	defer client.Close(context.Background())

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	var mu sync.Mutex
	var received []*Message
	err = client.Queue("q1").Wait(true).Limit(1).AutoAck(false).Consume(ctx, func(_ context.Context, msg *Message) error {
		mu.Lock()
		received = append(received, msg)
		mu.Unlock()
		return nil
	}).Execute(ctx)
	if err != nil {
		t.Fatalf("Consume failed: %v", err)
	}
	if len(received) != 1 || received[0].TransactionID != "tx-1" {
		t.Errorf("received = %+v, want a single tx-1 message", received)
	}
	if hitCount(hits) < 4 {
		t.Errorf("hits = %d, want >= 4 (3 rate-limited attempts + the final success)", hitCount(hits))
	}
}

func TestConsumeStopsOnTerminal403(t *testing.T) {
	url, hits := newPlanServer(t, nil, planResponse{status: 403, body: map[string]interface{}{"error": "cluster suspended", "code": "cluster_suspended"}})
	client, err := New(ClientConfig{URL: url})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	defer client.Close(context.Background())

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	err = client.Queue("q1").Wait(true).Consume(ctx, func(_ context.Context, _ *Message) error {
		return nil
	}).Execute(ctx)

	if err == nil {
		t.Fatal("expected Consume to stop with an error")
	}
	var httpErr *HTTPError
	if !errors.As(err, &httpErr) || !httpErr.IsClusterSuspended() {
		t.Errorf("err = %v, want a cluster_suspended HTTPError", err)
	}
	if hitCount(hits) != 1 {
		t.Errorf("hits = %d, want 1 -- must stop after the first 403, not hot-loop", hitCount(hits))
	}
}

File: clients/client-js/test-v2/http-unit/retry429.test.js
/**
 * Client-side 429/403 handling tests (PLAN_QUEEN_PROXY_CLOUD.md §4/§9,
 * blocker B4 -- "client 429/backoff work ... mandatory pre-enforcement").
 *
 * The proxy error contract under test:
 *   429  Retry-After: <seconds>  { "error": "...", "code": "rate_limited" | "quota_exceeded" }
 *   403                          { "error": "...", "code": "cluster_suspended" | "storage_quota_exceeded"
 *                                                          | "feature_gated" | "forbidden" }
 *
 * These tests run against a real local HTTP server (no broker) that plays
 * back a canned response script, mirroring the style of ack.test.js
 * (withAckServer) and streams-unit/fakeServer.js -- no mocking framework,
 * just node:http.
 */

import { describe, it } from 'node:test'
import assert from 'node:assert/strict'
import { createServer } from 'node:http'

import { Queen } from '../../client-v2/index.js'
import { HttpClient } from '../../client-v2/http/HttpClient.js'

/**
 * Spin up a real HTTP server driven by a canned response `plan` (array of
 * {status, body, retryAfter, headers}, consumed in request order; any
 * request beyond the plan gets `defaultResponse`). Records each hit's
 * arrival time so tests can sanity-check backoff pacing.
 */
function withPlanServer(plan, defaultResponse, run) {
  const hits = []
  const start = Date.now()
  let index = 0

  const server = createServer((req, res) => {
    let raw = ''
    req.on('data', chunk => { raw += chunk })
    req.on('end', () => {
      hits.push({ method: req.method, url: req.url, atMs: Date.now() - start })

      const descriptor = index < plan.length ? plan[index] : defaultResponse
      index++

      const status = descriptor.status ?? 200
      const headers = { 'Content-Type': 'application/json', ...(descriptor.headers || {}) }
      if (descriptor.retryAfter != null) headers['Retry-After'] = String(descriptor.retryAfter)

      res.writeHead(status, headers)
      res.end(JSON.stringify(descriptor.body ?? { ok: true }))
    })
  })

  return new Promise((resolve, reject) => {
    server.listen(0, '127.0.0.1', () => {
      const { port } = server.address()
      const teardown = () => new Promise(r => server.close(r))
      Promise.resolve()
        .then(() => run(`http://127.0.0.1:${port}`, hits))
        .then(
          async (value) => { await teardown(); resolve(value) },
          async (err) => { await teardown(); reject(err) }
        )
    })
  })
}

function rateLimited(retryAfter) {
  const descriptor = { status: 429, body: { error: 'slow down', code: 'rate_limited' } }
  if (retryAfter != null) descriptor.retryAfter = retryAfter
  return descriptor
}

function repeat(descriptor, n) {
  return Array.from({ length: n }, () => ({ ...descriptor }))
}

// ---------------------------------------------------------------------------
// HttpClient (direct) -- the centralized retry429 mechanism itself.
// ---------------------------------------------------------------------------

describe('HttpClient — 429 retry policy', () => {
  it('honors Retry-After (seconds) and succeeds once the header-specified wait elapses', async () => {
    const plan = [rateLimited('0')]
    await withPlanServer(plan, { status: 200, body: { ok: true } }, async (url, hits) => {
      const client = new HttpClient({ baseUrl: url, retry429: { baseMs: 5, capMs: 50 } })
      try {
        const result = await client.get('/x')
        assert.deepEqual(result, { ok: true })
        assert.equal(hits.length, 2, 'one 429 then one success')
      } finally {
        await client.destroy()
      }
    })
  })

  it('falls back to exponential backoff when Retry-After is absent, and the gap grows', async () => {
    const plan = repeat(rateLimited(), 2)
    await withPlanServer(plan, { status: 200, body: { ok: true } }, async (url, hits) => {
      const client = new HttpClient({ baseUrl: url, retry429: { baseMs: 20, capMs: 2000 } })
      try {
        const result = await client.get('/x')
        assert.deepEqual(result, { ok: true })
        assert.equal(hits.length, 3)
        const gap1 = hits[1].atMs - hits[0].atMs
        const gap2 = hits[2].atMs - hits[1].atMs
        // Exponential (base=20ms -> ~20ms then ~40ms), jittered +-20%: allow
        // generous slack for CI timer jitter while still proving growth.
        assert.ok(gap2 > gap1 * 1.2, `expected gap2 (${gap2}ms) to exceed gap1 (${gap1}ms) by backoff growth`)
      } finally {
        await client.destroy()
      }
    })
  })

  it('gives up after retry429.maxAttempts on a default (push-like) call and throws rate_limited', async () => {
    await withPlanServer([], rateLimited(), async (url, hits) => {
      const client = new HttpClient({ baseUrl: url, retry429: { maxAttempts: 3, baseMs: 1, capMs: 5 } })
      try {
        await assert.rejects(
          client.post('/api/v1/push', { items: [] }),
          (err) => {
            assert.equal(err.status, 429)
            assert.equal(err.code, 'rate_limited')
            return true
          }
        )
        assert.equal(hits.length, 3, 'exactly maxAttempts tries, no more')
      } finally {
        await client.destroy()
      }
    })
  })

  it('defaults to 10 attempts for a default (non-pop) call when retry429 is not configured', async () => {
    await withPlanServer([], rateLimited(), async (url, hits) => {
      const client = new HttpClient({ baseUrl: url, retry429: { baseMs: 1, capMs: 2 } })
      try {
        await assert.rejects(client.post('/api/v1/push', { items: [] }))
        assert.equal(hits.length, 10)
      } finally {
        await client.destroy()
      }
    })
  })

  it('retries a long-poll pop (retryKind "pop") past the push default of 10 attempts', async () => {
    const plan = repeat(rateLimited('0'), 14)
    await withPlanServer(plan, { status: 200, body: { messages: [{ id: 'm1' }] } }, async (url, hits) => {
      const client = new HttpClient({ baseUrl: url, retry429: { baseMs: 1, capMs: 5 } })
      try {
        const result = await client.get('/api/v1/pop', null, null, 'pop')
        assert.deepEqual(result, { messages: [{ id: 'm1' }] })
        assert.equal(hits.length, 15, 'pop must not give up at the push default of 10 attempts')
      } finally {
        await client.destroy()
      }
    })
  })

  it('an explicit maxAttempts override applies to pop too (not just push)', async () => {
    await withPlanServer([], rateLimited(), async (url, hits) => {
      const client = new HttpClient({ baseUrl: url, retry429: { maxAttempts: 2, baseMs: 1, capMs: 5 } })
      try {
        await assert.rejects(client.get('/api/v1/pop', null, null, 'pop'))
        assert.equal(hits.length, 2, 'explicit maxAttempts bounds pop as well as push')
      } finally {
        await client.destroy()
      }
    })
  })

  it('never retries a 403 and preserves its code (terminal)', async () => {
    await withPlanServer([], { status: 403, body: { error: 'cluster suspended', code: 'cluster_suspended' } }, async (url, hits) => {
      const client = new HttpClient({ baseUrl: url })
      try {
        await assert.rejects(
          client.post('/api/v1/push', { items: [] }),
          (err) => {
            assert.equal(err.status, 403)
            assert.equal(err.code, 'cluster_suspended')
            return true
          }
        )
        assert.equal(hits.length, 1, '403 must not be retried')
      } finally {
        await client.destroy()
      }
    })
  })

  it('still does not retry an ordinary 4xx like 400 (unaffected by the 429 changes)', async () => {
    await withPlanServer([], { status: 400, body: { error: 'bad request' } }, async (url, hits) => {
      const client = new HttpClient({ baseUrl: url })
      try {
        await assert.rejects(client.get('/x'), (err) => {
          assert.equal(err.status, 400)
          return true
        })
        assert.equal(hits.length, 1)
      } finally {
        await client.destroy()
      }
    })
  })
})

// ---------------------------------------------------------------------------
// Queen public API wiring: config plumbing (retry429) + call-site marking
// (push vs. wait=true pop) actually reach HttpClient.
// ---------------------------------------------------------------------------

describe('Queen — push()/pop() wired to retry429', () => {
  it('push() retries a 429 (Retry-After) then succeeds', async () => {
    const plan = [rateLimited('0')]
    await withPlanServer(plan, { status: 200, body: [{ status: 'queued', transactionId: 'tx-1' }] }, async (url, hits) => {
      const queen = new Queen({ url, handleSignals: false, retry429: { baseMs: 5, capMs: 50 } })
      try {
        const result = await queen.queue('q1').push({ hello: 'world' })
        assert.equal(hits.length, 2)
        assert.equal(result[0].status, 'queued')
      } finally {
        await queen.close()
      }
    })
  })

  it('push() surfaces a terminal 403 without retrying', async () => {
    await withPlanServer([], { status: 403, body: { error: 'over quota', code: 'storage_quota_exceeded' } }, async (url, hits) => {
      const queen = new Queen({ url, handleSignals: false })
      try {
        await assert.rejects(
          async () => { await queen.queue('q1').push({ hello: 'world' }) },
          (err) => { assert.equal(err.code, 'storage_quota_exceeded'); return true }
        )
        assert.equal(hits.length, 1)
      } finally {
        await queen.close()
      }
    })
  })

  it('a long-poll pop() rides out more 429s than the push default before returning', async () => {
    const plan = repeat(rateLimited('0'), 12)
    await withPlanServer(plan, { status: 200, body: { messages: [{ transactionId: 'tx-1', data: { x: 1 } }] } }, async (url, hits) => {
      const queen = new Queen({ url, handleSignals: false, retry429: { baseMs: 1, capMs: 5 } })
      try {
        const messages = await queen.queue('q1').wait(true).pop()
        assert.equal(messages.length, 1)
        assert.equal(hits.length, 13)
      } finally {
        await queen.close()
      }
    })
  })

  it('a non-waiting pop() uses the bounded default and swallows an exhausted 429 to []', async () => {
    await withPlanServer([], rateLimited(), async (url, hits) => {
      const queen = new Queen({ url, handleSignals: false, retry429: { maxAttempts: 2, baseMs: 1, capMs: 5 } })
      try {
        const messages = await queen.queue('q1').wait(false).pop()
        assert.deepEqual(messages, [], 'pop() keeps its swallow-errors-to-[] contract')
        assert.equal(hits.length, 2)
      } finally {
        await queen.close()
      }
    })
  })
})

// ---------------------------------------------------------------------------
// ConsumerManager worker loop: the actual hot-loop/die bug (B4) this task
// fixes -- consume() must back off through 429s and stop cleanly on a
// terminal 403 instead of spinning or crashing uncontrolled.
// ---------------------------------------------------------------------------

describe('Queen — consume() loop backs off on 429 and stops on terminal 403', () => {
  it('backs off through repeated 429s and still delivers the message', async () => {
    const plan = repeat(rateLimited('0'), 3)
    await withPlanServer(
      plan,
      { status: 200, body: { messages: [{ transactionId: 'tx-1', partitionId: 'p-1', data: { x: 1 } }] } },
      async (url, hits) => {
        const queen = new Queen({ url, handleSignals: false, retry429: { baseMs: 1, capMs: 5 } })
        try {
          const received = []
          // .each() delivers one message per handler call (default batches
          // the whole array) -- simplest shape to assert on here.
          await queen.queue('q1').wait(true).limit(1).autoAck(false).each().consume(async (msg) => {
            received.push(msg)
          })
          assert.equal(received.length, 1)
          assert.equal(received[0].transactionId, 'tx-1')
          assert.ok(hits.length >= 4, 'the 3 rate-limited attempts plus the final success must all have hit the server')
        } finally {
          await queen.close()
        }
      }
    )
  })

  it('stops the worker on a terminal 403 (cluster_suspended) instead of hot-looping or hanging', async () => {
    await withPlanServer([], { status: 403, body: { error: 'cluster suspended', code: 'cluster_suspended' } }, async (url, hits) => {
      const queen = new Queen({ url, handleSignals: false })
      try {
        await assert.rejects(
          async () => { await queen.queue('q1').wait(true).consume(async () => {}) },
          (err) => { assert.equal(err.code, 'cluster_suspended'); return true }
        )
        assert.equal(hits.length, 1, 'must stop after the first 403, not hot-loop')
      } finally {
        await queen.close()
      }
    })
  })
})
